rest/utilities_testing.go

```diff
@@ -498,39 +505,131 @@ func (s *SlowResponseRecorder) Write(buf []byte) (int, error) {
+	// If a RestTester is passed in, certain properties of the BlipTester such as noAdminParty will be ignored, since
+	// those properties only affect the creation of the RestTester.
+	// If nil, a default restTester will be created based on the properties in this spec
+	restTester *RestTester
+}
+
+// State associated with a BlipTester
 type BlipTester struct {
-	rt          RestTester
-	sender      *blip.Sender
+
+	// The underlying RestTester which is used to bootstrap the initial blip websocket creation,
+	// as well as providing a way for tests to access Sync Gateway over REST to hit admin-only endpoints
+	// which are not available via blip.  Since a test may need to create multiple BlipTesters for multiple
+	// user contexts, a single RestTester may be shared among multiple BlipTester instances.
+	restTester *RestTester
+
+	// The blip context which contains blip related state and the sender/reciever goroutines associated
+	// with this websocket connection
 	blipContext *blip.Context
+
+	// The blip sender that can be used for sending messages over the websocket connection
+	sender *blip.Sender
 }
 
+// Close the bliptester
 func (bt BlipTester) Close() {
-	bt.rt.Close()
+	bt.restTester.Close()
+}
+
+// Create a BlipTester using the default spec
+func NewBlipTester() *BlipTester {
+	defaultSpec := BlipTesterSpec{}
+	return NewBlipTesterFromSpec(defaultSpec)
 }
 
-func CreateBlipTester(t *testing.T, noConflictsMode bool) (bt BlipTester) {
+// Create a BlipTester using the given spec
+func NewBlipTesterFromSpec(spec BlipTesterSpec) *BlipTester {
 
 	EnableBlipSyncLogs()
 
-	// Create an admin handler
-	var adminHandler http.Handler
-	switch noConflictsMode {
-	case true:
-		adminHandler = bt.rt.TestAdminHandlerNoConflictsMode()
-	default:
-		adminHandler = bt.rt.TestAdminHandler()
+	bt := &BlipTester{}
 
+	if spec.restTester != nil {
+		bt.restTester = spec.restTester
+	} else {
+		rt := RestTester{
+			EnableNoConflictsMode: spec.noConflictsMode,
+			noAdminParty:          spec.noAdminParty,
+		}
+		bt.restTester = &rt
 	}
 
-	// Create a test server and close it when the test is complete
-	srv := httptest.NewServer(adminHandler)
+	// Since blip requests all go over the public handler, wrap the public handler with the httptest server
+	publicHandler := bt.restTester.TestPublicHandler()
+
+	if len(spec.connectingUsername) > 0 {
+
+		// By default, the user will be granted access to a single channel equal to their username
+		adminChannels := []string{spec.connectingUsername}
+
+		// If the caller specified a list of channels to grant the user access to, then use that instead.
+		if len(spec.connectingUserChannelGrants) > 0 {
+			adminChannels = []string{} // empty it
+			adminChannels = append(adminChannels, spec.connectingUserChannelGrants...)
+		}
+
+		// serialize admin channels to json array
+		adminChannelsJson, err := json.Marshal(adminChannels)
+		if err != nil {
+			panic(fmt.Sprintf("Error marshalling adming channels: %v", err))
```

Comment:
I wasn't sure about panic vs. returning error here (and below)- is the idea that since NewBlipTesterFromSpec is usually called during test setup, there's no harm in just panicking?  I suppose that's reasonable (and avoids having to do an error check in every caller), but feels a bit non-idiomatic.

Reply:
Ok, originally I had `bt, err := ..; if err != nil {}` everywhere, which made the code harder to follow.  But using `assertNoError()` on a one-liner makes things easier to read, so I switched to that.  Updated in https://github.com/couchbase/sync_gateway/pull/3217/commits/306425ac866224535180fb154d297d61695f6712